src/codegen/operator/index_scan_translator.cpp

```diff
@@ -0,0 +1,343 @@
+
+  // the iterator makes function call to the index
+  codegen.Call(IndexScanIteratorProxy::DoScan, {iterator_ptr});
+
+  const uint32_t num_columns =
+      static_cast<uint32_t>(table.GetSchema()->GetColumnCount());
+  llvm::Value *column_layouts = codegen->CreateAlloca(
+      ColumnLayoutInfoProxy::GetType(codegen), codegen.Const32(num_columns));
+
+  // get the index scan result size and iterator over all the results
+  llvm::Value *result_size =
+      codegen.Call(IndexScanIteratorProxy::GetResultSize, {iterator_ptr});
+  llvm::Value *result_iter = codegen.Const64(0);
+  lang::Loop loop{codegen,
+                  codegen->CreateICmpULT(result_iter, result_size),
+                  {{"distinctTileGroupIter", result_iter}}};
+  {
+    result_iter = loop.GetLoopVar(0);
+
+    // get pointer to the target tile group
+    llvm::Value *tile_group_id = codegen.Call(
+        IndexScanIteratorProxy::GetTileGroupId, {iterator_ptr, result_iter});
+    llvm::Value *tile_group_offset =
+        codegen.Call(IndexScanIteratorProxy::GetTileGroupOffset,
+                     {iterator_ptr, result_iter});
+    llvm::Value *tile_group_ptr =
+        codegen.Call(RuntimeFunctionsProxy::GetTileGroupByGlobalId,
+                     {table_ptr, tile_group_id});
+
+    codegen.Call(
+        RuntimeFunctionsProxy::GetTileGroupLayout,
+        {tile_group_ptr, column_layouts, codegen.Const32(num_columns)});
+    // Collect <start, stride, is_columnar> triplets of all columns
+    std::vector<TileGroup::ColumnLayout> col_layouts;
+    auto *layout_type = ColumnLayoutInfoProxy::GetType(codegen);
+    for (uint32_t col_id = 0; col_id < num_columns; col_id++) {
+      auto *start = codegen->CreateLoad(codegen->CreateConstInBoundsGEP2_32(
+          layout_type, column_layouts, col_id, 0));
+      auto *stride = codegen->CreateLoad(codegen->CreateConstInBoundsGEP2_32(
+          layout_type, column_layouts, col_id, 1));
+      auto *columnar = codegen->CreateLoad(codegen->CreateConstInBoundsGEP2_32(
+          layout_type, column_layouts, col_id, 2));
+      col_layouts.push_back(
+          TileGroup::ColumnLayout{col_id, start, stride, columnar});
+    }
+
+    TileGroup tileGroup(*table.GetSchema());
+    TileGroup::TileGroupAccess tile_group_access{tileGroup, col_layouts};
+
+    // visibility
+    llvm::Value *executor_context_ptr =
+        this->GetCompilationContext().GetExecutorContextPtr();
+    llvm::Value *txn = codegen.Call(ExecutorContextProxy::GetTransaction,
+                                    {executor_context_ptr});
+    llvm::Value *raw_sel_vec = sel_vec.GetVectorPtr();
+    // Invoke TransactionRuntime::PerformRead(...)
+    llvm::Value *out_idx =
+        codegen.Call(TransactionRuntimeProxy::PerformVectorizedRead,
+                     {txn, tile_group_ptr, tile_group_offset,
+                      codegen->CreateAdd(tile_group_offset, codegen.Const32(1)),
+                      raw_sel_vec});
+    sel_vec.SetNumElements(out_idx);
```

Comment:
A lot of this code is duplicated from `codegen::TileGroup` and `codegen::Table`. Can you try to integrate it more cleanly?